File: apps/backend/app/schemas/unit.py
from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field

from app.models.unit import UtilityOverride


# ── Update ────────────────────────────────────────────────────────────────────

class UnitUpdateRequest(BaseModel):
    rent_base: Optional[float] = Field(default=None, ge=0)
    deposit_amount: Optional[float] = Field(default=None, ge=0)
    deposit_rule: Optional[str] = None
    utility_deposit: Optional[float] = Field(default=None, ge=0)
    status: Optional[str] = None  # vacant | reserved | occupied | inactive
    unit_type: Optional[str] = None
    size: Optional[float] = Field(default=None, ge=0)
    furnished: Optional[bool] = None
    is_premium: Optional[bool] = None
    utility_overrides: Optional[UtilityOverride] = None
    meter_id: Optional[str] = None
    iot_device_id: Optional[str] = None


class BulkUnitUpdate(BaseModel):
    unit_id: str
    updates: UnitUpdateRequest


class BulkUpdateRequest(BaseModel):
    updates: List[BulkUnitUpdate] = Field(min_length=1, max_length=500)


class BulkUpdateResponse(BaseModel):
    updated: int
    failed: int
    errors: List[Dict[str, Any]] = []


# ── Response ──────────────────────────────────────────────────────────────────

class UnitResponse(BaseModel):
    id: str
    org_id: str
    property_id: str
    unit_code: str
    wing: Optional[str]
    floor: int
    unit_number: str
    unit_type: str
    size: Optional[float]
    furnished: bool
    is_premium: bool
    status: str
    rent_base: Optional[float]
    deposit_amount: Optional[float]
    deposit_rule: Optional[str]
    utility_deposit: Optional[float]
    utility_overrides: Optional[UtilityOverride]
    meter_id: Optional[str]
    iot_device_id: Optional[str]
    created_at: datetime
    updated_at: datetime


class UnitListResponse(BaseModel):
    items: List[UnitResponse]
    total: int
    page: int
    page_size: int


# ── Reserve ───────────────────────────────────────────────────────────────────

class UnitReserveRequest(BaseModel):
    tenant_id: str
    onboarding_id: Optional[str] = None

File: apps/frontend/src/types/unit.ts
import type { PricingTier } from '@/types/property'
export type { PricingTier }

export type UnitStatus = 'vacant' | 'reserved' | 'occupied' | 'inactive'

export interface UtilityOverrideDetail {
  type: 'shared' | 'metered' | 'subscription'
  rate?: number
  unit?: string
  label?: string
  income_account?: string  // chart-of-accounts code override for this unit
  tiers?: PricingTier[]    // tiered pricing when type='metered'
  deposit?: number         // per-utility deposit override for this unit
}

export interface UtilityOverride {
  electricity?: UtilityOverrideDetail
  water?: UtilityOverrideDetail
  gas?: UtilityOverrideDetail
  internet?: UtilityOverrideDetail
  garbage?: UtilityOverrideDetail
  security?: UtilityOverrideDetail
}

export interface MeterReadingCacheEntry {
  value: number
  read_at: string       // ISO datetime
  read_by: string       // user_id
  read_by_name: string  // resolved display name
}

export interface Unit {
  id: string
  org_id: string
  property_id: string
  unit_code: string
  wing?: string
  floor: number
  unit_number: string
  unit_type: string
  size?: number
  furnished: boolean
  is_premium: boolean
  status: UnitStatus
  rent_base?: number
  deposit_amount?: number
  deposit_rule?: string
  utility_deposit?: number
  utility_overrides?: UtilityOverride
  meter_reading_cache?: Record<string, MeterReadingCacheEntry>
  meter_id?: string
  iot_device_id?: string
  created_at: string
  updated_at: string
}

export interface UnitUpdateRequest {
  rent_base?: number
  deposit_amount?: number
  deposit_rule?: string
  utility_deposit?: number
  status?: UnitStatus
  unit_type?: string
  size?: number
  furnished?: boolean
  is_premium?: boolean
  utility_overrides?: UtilityOverride
  meter_id?: string
  iot_device_id?: string
}

export interface UtilityLineItem {
  key: string
  label: string
  type: string
  rate?: number
  unit_label?: string
  income_account?: string
  deposit?: number
}

export interface UnitPricingResponse {
  unit_id: string
  unit_code: string
  rent_amount: number
  deposit_amount: number
  deposit_rule: string
  utility_deposit?: number
  utilities: UtilityLineItem[]
  prorated_rent: number
  prorated_days: number
  days_in_month: number
  total_move_in: number
}

export interface BulkUnitUpdate {
  unit_id: string
  updates: UnitUpdateRequest
}

export interface BulkUpdateRequest {
  updates: BulkUnitUpdate[]
}

export interface BulkUpdateResponse {
  updated: number
  failed: number
  errors: Array<{ unit_id: string; error: string }>
}
